theano/tensor/nnet/abstract_conv2d.py

```diff
@@ -0,0 +1,741 @@
+"""
+FIXME
+"""
+
+__docformat__ = "restructuredtext en"
+
+import logging
+
+import numpy
+
+import theano
+from theano.tensor import (as_tensor_variable, blas, get_scalar_constant_value,
+                           patternbroadcast, NotScalarConstantError)
+from theano.tensor import TensorType
+from theano.gof import Apply, Op
+from theano.gof import local_optimizer
+
+from theano.sandbox.cuda import register_opt as register_gpu
+from theano.tensor.opt import register_specialize_device
+
+
+### Gpu related optimization (to be moved in sandbox/cuda)
+from theano.sandbox.cuda.basic_ops import (
+    as_cuda_ndarray_variable,
+    gpu_contiguous, gpu_from_host, host_from_gpu,
+    GpuFromHost, HostFromGpu
+    )
+from theano.sandbox.cuda.type import CudaNdarrayType
+from theano.sandbox.cuda.dnn import dnn_available, dnn_conv, dnn_gradweight, dnn_gradinput
+from theano.sandbox.cuda.blas import GpuCorrMM, GpuCorrMM_gradWeights, GpuCorrMM_gradInputs
+from theano.sandbox.cuda.opt import values_eq_approx_high_tol
+
+
+## Cpu implementation
+from theano.tensor.nnet import conv2d as cpu_conv2d, ConvOp
+from theano.tensor.nnet.ConvGrad3D import convGrad3D
+from theano.tensor.nnet.ConvTransp3D import convTransp3D
+
+_logger = logging.getLogger("theano.tensor.nnet.conv2d")
+
+
+def conv2d(inputs,
+           filters,
+           inputs_shape=None,
+           filters_shape=None,
+           batch_size=None,
+           border_mode='valid',
+           subsample=(1, 1),
+           filters_flip=True):
+    """
+    This function will build the symbolic graph for convolving a mini-batch of a
+    stack of 2D inputs with a set of 2D filters. The implementation is modelled
+    after Convolutional Neural Networks (CNN).
+
+    :type inputs: symbolic 4D tensor
+    :param input: mini-batch of feature map stacks, of shape
+        (batch size, input channels, input rows, input columns).
+        See the optional parameter ``input_shape``.
+
+    :type filters: symbolic 4D tensor
+    :param filters: set of filters used in CNN layer of shape
+        (output channels, input channels, filter rows, filter columns).
+        See the optional parameter ``filter_shape``.
+
+    :type inputs_shape: None, tuple/list of len 4 of int or Constant variable
+    :param input_shape: The shape of the input parameter.
+        Optional, possibly used to choose an optimal implementation.
+        You can give ``None`` for any element of the list to specify that this
+        element is not known at compile time.
+
+    :type filters_shape: None, tuple/list of len 4 of int or Constant variable
+    :param filter_shape: The shape of the filters parameter.
```

Comment:
There is some inconsistency `input_shape` / `inputs_shape`, `filter_shape` / `filters_shape`. It was called `input_shape` and `filter_shape` before, it's probably not a good idea to change it.